crates/noirc_driver/src/lib.rs

```diff
@@ -169,18 +166,46 @@ impl Driver {
         }
     }
 
-    /// Run the lexing, parsing, name resolution, and type checking passes,
-    /// returning Err(FrontendError) and printing any errors that were found.
-    pub fn check_crate(&mut self) -> Result<(), Vec<FileDiagnostic>> {
-        let mut errs = vec![];
-        CrateDefMap::collect_defs(LOCAL_CRATE, &mut self.context, &mut errs);
-        if errs.is_empty() {
-            Ok(())
+    /// Run the lexing, parsing, name resolution, and type checking passes.
+    ///
+    /// This returns a (possibly empty) vector of any warnings found on success.
+    /// On error, this returns a non-empty vector of warnings and error messages, with at least one error.
+    pub fn check_crate(&mut self, deny_warnings: bool) -> Result<Warnings, ErrorsAndWarnings> {
+        let mut errors = vec![];
+        CrateDefMap::collect_defs(LOCAL_CRATE, &mut self.context, &mut errors);
+
+        if Self::has_errors(&errors, deny_warnings) {
+            Err(errors)
         } else {
-            Err(errs)
+            Ok(errors)
         }
     }
 
+    /// Run the lexing, parsing, name resolution, and type checking passes and report any warnings
+    /// and errors found.
+    pub fn check_crate_and_report_errors(
+        &mut self,
+        deny_warnings: bool,
+    ) -> Result<(), ReportedErrors> {
+        let result = self.check_crate(deny_warnings).map(|warnings| ((), warnings));
+        self.report_errors(result, deny_warnings)
+    }
+
+    /// Helper function for reporting any errors in a Result<(T, Warnings), ErrorsAndWarnings>
+    /// structure that is commonly used as a return result in this file.
+    pub fn report_errors<T>(
+        &self,
+        result: Result<(T, Warnings), ErrorsAndWarnings>,
+        deny_warnings: bool,
+    ) -> Result<T, ReportedErrors> {
+        let (t, warnings) = result.map_err(|errors| {
+            noirc_errors::reporter::report_all(self.file_manager(), &errors, deny_warnings)
+        })?;
+
+        noirc_errors::reporter::report_all(self.file_manager(), &warnings, deny_warnings);
+        Ok(t)
+    }
```

Comment:
We'll want this inside of nargo_cli, as this should be pure from stdout/stderr stuff. We can add it as a utility over there.